File: java/org.financer.client/src/main/java/org/financer/client/domain/model/transaction/VariableTransaction.java
package org.financer.client.domain.model.transaction;

import org.financer.client.domain.model.category.Category;
import org.financer.shared.domain.model.AmountProvider;
import org.financer.shared.domain.model.value.objects.Amount;
import org.financer.shared.domain.model.value.objects.TimeRange;
import org.financer.shared.domain.model.value.objects.ValueDate;

import java.util.HashSet;
import java.util.Set;

public final class VariableTransaction extends Transaction {

    private ValueDate valueDate;
    private Set<Product> products = new HashSet<>();

    @Override
    public Amount getAmount() {
        Amount amount = new Amount();
        for (AmountProvider amountProvider : this.products) {
            amount = amount.add(amountProvider.getAmount());
        }
        return amount;
    }

    @Override
    public Amount getAmount(ValueDate valueDate) {
        Amount amount = new Amount();
        for (AmountProvider amountProvider : this.products) {
            amount = amount.add(amountProvider.getAmount(valueDate));
        }
        return amount;
    }

    @Override
    public Amount getAmount(TimeRange timeRange) {
        Amount amount = new Amount();
        for (AmountProvider amountProvider : this.products) {
            amount = amount.add(amountProvider.getAmount(timeRange));
        }
        return amount;
    }

    @Override
    public void adjustAmountSign() {
        for (AmountProvider amountProvider : this.getProducts()) {
            amountProvider.adjustAmountSign();
        }
    }

    /*
     * Getters and Setters
     */

    public ValueDate getValueDate() {
        return valueDate;
    }

    public VariableTransaction setValueDate(ValueDate valueDate) {
        this.valueDate = valueDate;
        return this;
    }

    public Set<Product> getProducts() {
        return products;
    }

    public VariableTransaction setProducts(Set<Product> products) {
        this.products = products;
        return this;
    }

    public VariableTransaction addProduct(Product product) {
        this.products.add(product);
        return this;
    }

    @Override
    public VariableTransaction setId(long id) {
        super.setId(id);
        return this;
    }

    @Override
    public VariableTransaction setCategory(Category category) {
        super.setCategory(category);
        return this;
    }

    @Override
    public VariableTransaction setDescription(String purpose) {
        super.setDescription(purpose);
        return this;
    }

    @Override
    public VariableTransaction setVendor(String vendor) {
        super.setVendor(vendor);
        return this;
    }

    @Override
    public VariableTransaction setAttachments(Set<Attachment> attachments) {
        super.setAttachments(attachments);
        return this;
    }
}

File: java/org.financer.client/src/main/kotlin/org/financer/client/domain/model/transaction/VariableTransaction.kt
package org.financer.client.domain.model.transaction

import org.financer.client.domain.model.category.Category
import org.financer.shared.domain.model.value.objects.Amount
import org.financer.shared.domain.model.value.objects.TimeRange
import org.financer.shared.domain.model.value.objects.ValueDate
import java.util.*

class VariableTransaction : Transaction() {
    /*
     * Getters and Setters
     */ var valueDate: ValueDate? = null
        private set
    private var products: MutableSet<Product> = HashSet()
    override fun getAmount(): Amount {
        var amount = Amount()
        for (amountProvider in products) {
            amount = amount.add(amountProvider.amount)
        }
        return amount
    }

    override fun getAmount(valueDate: ValueDate): Amount {
        var amount = Amount()
        for (amountProvider in products) {
            amount = amount.add(amountProvider.getAmount(valueDate))
        }
        return amount
    }

    override fun getAmount(timeRange: TimeRange): Amount {
        var amount = Amount()
        for (amountProvider in products) {
            amount = amount.add(amountProvider.getAmount(timeRange))
        }
        return amount
    }

    override fun adjustAmountSign() {
        for (amountProvider in getProducts()) {
            amountProvider.adjustAmountSign()
        }
    }

    fun setValueDate(valueDate: ValueDate?): VariableTransaction {
        this.valueDate = valueDate
        return this
    }

    fun getProducts(): Set<Product> {
        return products
    }

    fun setProducts(products: MutableSet<Product>): VariableTransaction {
        this.products = products
        return this
    }

    fun addProduct(product: Product): VariableTransaction {
        products.add(product)
        return this
    }

    override fun setId(id: Long): VariableTransaction {
        super.setId(id)
        return this
    }

    override fun setCategory(category: Category?): VariableTransaction {
        super.setCategory(category)
        return this
    }

    override fun setDescription(purpose: String?): VariableTransaction {
        super.setDescription(purpose)
        return this
    }

    override fun setVendor(vendor: String?): VariableTransaction {
        super.setVendor(vendor)
        return this
    }

    override fun setAttachments(attachments: MutableSet<Attachment>): VariableTransaction {
        super.setAttachments(attachments)
        return this
    }
}